compiler/src/lnntoamm/Statement.ts

```diff
@@ -0,0 +1,785 @@
+        TODO('object literals');
+      } else if (work.has('functions')) {
+        TODO('functions in functions');
+      } else if (work.has('variable')) {
+        const varName = work.get('variable').t;
+        if (ii === asts.length - 1) {
+          let dec = metadata.var(varName);
+          if (dec === null) {
+            throw new Error(`${varName} not defined`);
+          }
+          stmts.push(dec.ref());
+          break;
+        }
+        const next = asts[ii + 1];
+        if (next.has('fncall')) {
+          // make things nice and pretty :)
+          let callAst = new NamedAnd(
+            work.get('variable').t + next.get('fncall').t,
+            {
+              fnname: work.get('variable'),
+              fncall: next.get('fncall'),
+            },
+            (work as NamedOr).filename,
+            work.line,
+            work.char,
+          );
+          stmts.push(...Call.fromAsts(
+            callAst,
+            // null,
+            varName,
+            next.get('fncall'),
+            metadata,
+          ));
+          ii += 1;
+        } else if (next.has('methodsep')) {
+          TODO('accesses/methods');
+        } else {
+          throw new Error(`unexpected token: expected dot or call, found ${next}`);
+        }
+      } else if (work.has('constants')) {
+        work = work.get('constants');
+        let dec = Dec.generate(Lit.fromConstantsAst(work, metadata));
+        metadata.define(dec);
+        stmts.push(dec);
+      } else {
+        // TODO: don't lump in HOF and chains
+        throw new Error(`unexpected token: expected variable or value, found ${work.t.trim()}`);
+      }
+    }
+
+    return stmts;
+  }
+}
+
+export class Assign extends Stmt {
+  upstream: Dec
+  val: Stmt
+
+  constructor(
+    ast: LPNode,
+    upstream: Dec,
+    assignTo: Stmt,
+  ) {
+    super(ast);
+    this.upstream = upstream;
+    this.val = assignTo;
+  }
+
+  static fromAssignmentsAst(ast: LPNode, metadata: MetaData): Stmt[] {
+    let stmts: Stmt[] = [];
+    const name = ast.get('varn').t;
+    const upstream = metadata.var(name);
+    if (upstream === null) {
+      throw new Error(`can't assign to ${name}: not found`);
+    }
+    stmts.push(...Stmt.fromAssignables(ast.get('assignables'), metadata));
+    const expr = stmts.pop();
+    let assign: Assign;
+    if (expr instanceof Dec) {
+      assign = new Assign(ast, upstream, expr.val);
+    } else if (expr instanceof Ref) {
+      TODO('figure out how to resolve this???');
+    } else {
+      throw new Error(`invalid assignment state: not a declaration`);
+    }
+    upstream.ty.constrain(assign.val.exprTy());
+    stmts.push(assign);
+    return stmts;
+  }
+
+  exprTy(): Type {
+    throw new Error(`assignments aren't expressions`);
+  }
+
+  inline(amm: Output) {
+    const name = this.upstream.ammName;
+    const ty = this.upstream.ty.breakdown(); // always use the declaration's type, since it's been reduced.
```

Comment:
what about something like:

```ln
let foo = none();
foo = some(5);
```

Using the two resolved types for constraints seems to be in order here? But this depends on generic types being in, so I can see why it's not here yet.

That's also not entirely academic, I could see someone doing:

```ln
let out = none();
if myCondition {
  out = some(3.14);
} else if myOtherCondition {
  out = some(6.28);
}
return out;
```

Reply:
that'll still work - the generated type for `foo` will initially be `Maybe<Generated>`, then the assignments will constrain the generated type to `oneOf([float32, float64, int8, int16, int32, int64])` in your first example and both reassignments will constrain it to `oneOf([float32, float64])` in your second example